### PR TITLE
Implement setindex_raw!, fq_nmod, coeffs_raw

### LINKED ISSUE #999: Add a `coeffs` method for `fq_nmod`
For somewhat efficient conversion between `fq_nmod` and GAP finite fields of larger degree, it would be useful to have a method `coeffs` for `fq_nmod` which returns a `Vector{UInt}` with all coefficient in one go, instead of having to get each coefficient on its own (which is substantially slower).

Going the other way around, of course a constructor for `fq_nmod` taking such a vector would also be useful.

In the end, I mostly care about matrices, though; I guess there an in-place way to set the coefficients of an entry of a  `fq_nmod_mat` would be useful (say by having a `setindex!(a::fq_nmod_mat, u::Vector{UInt}, i::Int, j::Int)` method -- though of course it could also have a different, more internal sounding name like `_setindex!`.

I guess this requires either new APIs in FLINT, or we must be willing to directly poke into the `coeffs` field of `nmod_poly_struct` (since it is the first one, no annoying alignment issues crop up, so it wouldn't be too hard, and I guess more efficient than a `ccall`, too)

### PR DESCRIPTION
**Pull request by wbhart**
Fixes #999

### CODE REVIEW

File: src/flint/fq_nmod.jl [hunk 1]
```diff
@@ -49,6 +49,44 @@ function coeff(x::fq_nmod, n::Int)
...
+      V[i + 1] = c
+      if i + 1 == x.length && iszero(c)
+         while i >= 0 && iszero(V[i + 1])
+            x.length -= 1
+            i -= 1
+         end
+      end
+   end
+   return x
+end
```
Comment (tthsqe12) [inline src/flint/fq_nmod.jl:88]: What is this function supposed to do?

File: src/flint/fq_nmod.jl [hunk 2]
```diff
@@ -49,6 +49,44 @@ function coeff(x::fq_nmod, n::Int)
...
+function setindex_raw!(x::fq_nmod, c::UInt, i::Int)
+   len = degree(parent(x))
+   V = unsafe_wrap(Vector{UInt}, reinterpret(Ptr{UInt}, x.coeffs), len)
+   if i + 1 > x.length
+      if !iszero(c)
+         for j = x.length + 1:i
+            V[j] = 0
+         end
+         V[i + 1] = c
+         x.length = i + 1
```
Comment (tthsqe12) [inline src/flint/fq_nmod.jl:76]: I don't think we can just change `x.length` here. This is a memory corruption.

Comment (tthsqe12) [reply src/flint/fq_nmod.jl:76]: However, at the end of the day, would it just be better to pun the fq_nmod_struct into a nmod_poly_struct and call the correct nmod_poly function? Why write the same code twice in julia?

Comment (wbhart) [reply src/flint/fq_nmod.jl:76]: This type is supposed to always allocate deg(R) coefficients when it is initialised, so it shouldn't be a corruption. A lot of code would be broken if that weren't the case.

File: src/flint/fq_nmod.jl (hunk 1 above)
Comment (wbhart) [reply src/flint/fq_nmod.jl:88]: Read the ticket. It is an internal function requested for Gap interfacing.

Comment (tthsqe12) [reply src/flint/fq_nmod.jl:88]: Fair enough, but the ticket doesn't mention the restrictions under which this function is safe. This function can easily crash an burn.

File: src/flint/fq_nmod.jl (hunk 2 above)
Comment (tthsqe12) [reply src/flint/fq_nmod.jl:76]: > This type is supposed to always allocate deg(R) coefficients when it is initialised,
This unfortunately is of no consequence. The fq_nmod you get back after calling a flint function on an fq_nmod can have any alloc even if it went in with at least deg(R) alloc. Code like this in nemo is just asking for trouble.
I would pun it to a nmod_poly, call the correct nmod_poly function, and avoid the incorrect code duplication.

Comment (wbhart) [reply src/flint/fq_nmod.jl:76]: I will see if I can pun it in Nemo.

File: src/flint/fq_nmod.jl (hunk 1 above)
Comment (wbhart) [reply src/flint/fq_nmod.jl:88]: I changed it.

**Comment by tthsqe12** [comment]
Sorry. I think there is one more thing here: the constructor for fq_nmod here technically needs some GC preserve on r across the ccall because julia could decide the return object is not used. What do you think @thofma? Better to be safe then debugging obscure bugs on some julia versions..

**Comment by thofma** [comment]
Agreed.

**Comment by wbhart** [comment]
It's impossible for r to disappear here, since r.length is set to a value that can only be computed after the loop, which relies on V existing.

**Comment by tthsqe12** [comment]
Alirght.

**Comment by thofma** [comment]
Sorry, needs a rebase.

**Pull request merged**